--- a/share/spack/setup-env.sh
+++ b/share/spack/setup-env.sh
@@ -358,7 +358,15 @@ if [ -z "${SPACK_SKIP_MODULES+x}" ]; then
             _spack_pathadd PATH "${_sp_module_bin}"
         fi;
     else
-        eval `spack --print-shell-vars sh`
+		# This allows us to speed up the shell integration massively because
+		# we do not need to invoke Python. This only works if we only have one
+		# OS/architecture combination, though.
+		if [ -f "${_sp_share_dir}/setup-env.vars" ]
+		then
+			. "${_sp_share_dir}/setup-env.vars"
+		else
+			eval `spack --print-shell-vars sh`
+		fi
     fi;
 
 
